src/layers/embeddings/qk_rotary_embeddings.rs

```diff
@@ -1,8 +1,98 @@
-use candle_core::{Device, IndexOp, Tensor};
+use candle_core::{IndexOp, Tensor};
+use candle_nn::VarBuilder;
 use snafu::{ensure, ResultExt, Snafu};
 
 use crate::kv_cache::KeyValueCache;
-use crate::layers::embeddings::{RotaryEmbeddings, RotaryEmbeddingsError};
+use crate::layers::embeddings::{RotaryEmbeddings, RotaryEmbeddingsConfig, RotaryEmbeddingsError};
+
+/// Configuration for query-key rotary embeddings.
+#[derive(Debug)]
+pub struct QueryKeyRotaryEmbeddingsConfig {
+    base: usize,
+    fraction: f32,
+    head_width: usize,
+    seq_len: usize,
+}
+
+impl QueryKeyRotaryEmbeddingsConfig {
+    /// Build query-key rotary embeddings
+    pub fn build(
+        &self,
+        vb: VarBuilder,
+    ) -> Result<QueryKeyRotaryEmbeddings, QueryKeyRotaryEmbeddingsError> {
+        ensure!(
+            (0f32..=1f32).contains(&self.fraction),
+            InvalidRangeSnafu {
+                fraction: self.fraction
+            }
+        );
+
+        // Truncate to get the width in case it is fractional.
+        let rotary_width = (self.fraction * self.head_width as f32) as usize;
+        let rotary_embeds = RotaryEmbeddingsConfig::default()
+            .base(self.base)
+            .seq_len(self.seq_len)
+            .width(rotary_width)
+            .build(vb)
+            .context(RotaryEmbeddingsSnafu)?;
+
+        Ok(QueryKeyRotaryEmbeddings {
+            head_width: self.head_width,
+            rotary_embeds,
+            rotary_width,
+        })
+    }
+
+    /// Set the base used for `theta`.
```

Comment:
```suggestion
    /// Base used for `theta`.
```
To keep it consistent with the builder methods elsewhere.